File: examples/python/listen.py
from __future__ import annotations

import os
import sys
from pathlib import Path

from sipp import (
    ContextRuntimeConfig,
    Endpoint,
    ModelPlacementConfig,
    NativeRuntimeConfig,
    SippClient,
    set_llama_log_quiet,
)

from _support import gpu_layers, int_env


def main() -> None:
    if len(sys.argv) != 4:
        raise SystemExit(
            "usage: python examples/python/listen.py "
            "<model.gguf> <projector.gguf> <audio>"
        )
    model_path, projector_path, audio_path = sys.argv[1:]
    set_llama_log_quiet(True)

    client = SippClient()
    model = client.models.add([model_path, projector_path])
    client.add("asr", Endpoint.local(model, runtime=runtime_config(8192)))
    response = client.listen(
        Path(audio_path).read_bytes(),
        language=os.getenv("SIPP_LANGUAGE"),
    ).result()
    print(response["text"].strip())


def runtime_config(context_size: int) -> NativeRuntimeConfig:
    return NativeRuntimeConfig(
        placement=ModelPlacementConfig(gpu_layers=gpu_layers()),
        context=ContextRuntimeConfig(
            n_ctx=int_env("SIPP_CONTEXT", context_size),
            n_threads=int_env("SIPP_THREADS"),
            n_threads_batch=int_env("SIPP_THREADS"),
            warmup=False,
        ),
    )


if __name__ == "__main__":
    main()

File: examples/node/listen.mjs
import { readFileSync } from 'node:fs';

import native from '../../lib/node/router.js';
import { gpuLayers, intEnv } from './_support.mjs';

const [modelPath, projectorPath, audioPath] = process.argv.slice(2);
if (!modelPath || !projectorPath || !audioPath) {
  throw new Error('usage: node examples/node/listen.mjs <model.gguf> <projector.gguf> <audio>');
}

const { Endpoint, SippClient, setLlamaLogQuiet } = native;
setLlamaLogQuiet(true);
const client = new SippClient();
const model = await client.models.add([modelPath, projectorPath]);
await client.add('asr', Endpoint.local(model, { runtime: runtimeConfig(8192) }));

const response = await client.listen({
  audio: readFileSync(audioPath),
  language: process.env.SIPP_LANGUAGE,
}).response;
console.log(response.text.trim());

function runtimeConfig(contextSize) {
  return {
    placement: { gpu_layers: gpuLayers() },
    context: {
      n_ctx: intEnv('SIPP_CONTEXT', contextSize),
      n_threads: intEnv('SIPP_THREADS'),
      n_threads_batch: intEnv('SIPP_THREADS'),
      warmup: false,
    },
  };
}
